## Code Review: PR #3100

**chore: upgrade reqwest to 0.13 and tokio-tungstenite to 0.28**

### Summary

This PR bumps `reqwest` from 0.12 to 0.13 and `tokio-tungstenite` from 0.26 to 0.28 across the workspace, updating feature flags to match upstream API changes. The change is focused (6 files, mostly Cargo.toml edits) with a clear commit message and test plan. However, it introduces a **critical cross-compilation failure** that blocks merge.

---

### Critical Issues (P0)

#### `aws-lc-sys` breaks musl cross-compilation

The `runner-metal-test` CI job **fails** with linker errors when cross-compiling for `aarch64-unknown-linux-musl`:

```
undefined reference to `__memcpy_chk'
undefined reference to `__fprintf_chk'
undefined reference to `__vsnprintf_chk'
```

**Root cause:** reqwest 0.13's `rustls` feature pulls in `rustls-platform-verifier`, which transitively depends on `aws-lc-rs` / `aws-lc-sys`. The `aws-lc-sys` C library fails to link against musl targets.

**This is exactly the problem the `reqeast` crate was created to solve.** The `reqeast` wrapper (at `crates/reqeast/`) uses `reqwest` with `rustls-no-provider` and manually installs the `ring` crypto provider, specifically to avoid `aws-lc-sys`. However, this PR changes `guest-agent`, `runner`, and `ably-subscriber` to depend on `reqwest` directly with the `rustls` feature, bypassing `reqeast`.

**Suggested fix (pick one):**
- **(a)** Switch `guest-agent`, `runner`, and `ably-subscriber` to depend on `reqeast` instead of `reqwest` directly
- **(b)** Use `reqwest` with `rustls-no-provider` feature and handle crypto provider installation in each crate
- **(c)** Configure workspace-level feature exclusion to prevent `aws-lc-rs` from being pulled in

---

### High Priority Issues (P1)

#### 1. Inconsistent TLS backend in `ably-subscriber`

In `crates/ably-subscriber/Cargo.toml`, `tokio-tungstenite` still uses `rustls-tls-native-roots` (which pulls in `rustls-native-certs`), while `reqwest` was changed to `rustls` (which uses `rustls-platform-verifier`). This creates a split where two HTTP/WebSocket libraries in the same crate use different certificate verification backends.

#### 2. Incomplete README update

`crates/README.md` documents the reqwest 0.13 TLS behavior change but does not mention that crates must use `reqeast` rather than `reqwest` directly to avoid `aws-lc-sys` musl cross-compilation issues.

---

### New Transitive Dependencies Introduced

| Dependency | Notes |
|---|---|
| `aws-lc-rs` 1.15.4 | **Problematic** -- musl cross-compilation failure |
| `aws-lc-sys` 0.37.1 | **Problematic** -- C library with musl linkage issues |
| `cmake` 0.1.57 | Build dep for aws-lc-sys |
| `dunce` 1.0.5 | Build dep for aws-lc-sys |
| `fs_extra` 1.3.0 | Build dep for aws-lc-sys |
| `jobserver` 0.1.34 | Build dep for cc |

---

### Positive Aspects

- Clean, focused change with minimal scope
- Good commit message following conventional commits format
- Clear PR description with explicit test plan
- Local tests pass (365 tests), `cargo clippy` clean
- The version bumps themselves correctly track upstream reqwest 0.13 feature renames

---

### Verdict

**Changes Requested** -- The `aws-lc-sys` cross-compilation failure is a CI blocker. The core issue is that the three crates bypass the `reqeast` wrapper which was specifically designed to avoid this problem.

---

*Full review details: `codereviews/20260217/commit-list-pr3100.md`*

*Reviewed by Claude Code*

--- a/crates/Cargo.toml
+++ b/crates/Cargo.toml
@@ -13,6 +13,7 @@ members = [
     "vsock-guest",
     "vsock-host",
     "vsock-proto",
+    "reqeast",
     "vsock-test",
 ]
 
@@ -23,6 +24,7 @@ edition = "2024"
 # Internal crates
 ably-subscriber = { path = "ably-subscriber" }
 guest-common = { path = "guest-common" }
+reqeast = { path = "reqeast" }
 sandbox = { path = "sandbox" }
 sandbox-fc = { path = "sandbox-fc" }
 vsock-guest = { path = "vsock-guest" }
@@ -51,11 +53,12 @@ tracing = "0.1"
 tracing-subscriber = "0.3"
 which = "8"
 ureq = "3"
-reqwest = { version = "0.12", default-features = false }
+reqwest = { version = "0.13", default-features = false }
+rustls = { version = "0.23", default-features = false, features = ["ring", "std", "tls12"] }
 hmac = "0.12"
 sha2 = "0.10"
 tokio-util = "0.7"
-tokio-tungstenite = { version = "0.26", default-features = false }
+tokio-tungstenite = { version = "0.28", default-features = false }
 rmp-serde = "1"
 rmpv = "1"
 futures-util = "0.3"

--- a/crates/README.md
+++ b/crates/README.md
@@ -19,6 +19,7 @@ This workspace contains Rust crates for the vm0 sandbox runtime — VM orchestra
 | **guest-download** | Downloads and extracts storage archives based on manifest — parallel downloads with retry logic |
 | **guest-mock-claude** | Mock Claude CLI for testing — executes bash commands and outputs Claude-compatible JSONL |
 | **ably-subscriber** | Ably Pub/Sub subscribe-only realtime client — WebSocket/MessagePack protocol with token auth and automatic reconnection |
+| **reqeast** | Thin reqwest wrapper — auto-installs `ring` TLS crypto provider, avoiding `aws-lc-sys` musl cross-compilation issues |
 
 ## Architecture
 
@@ -44,7 +45,9 @@ This workspace contains Rust crates for the vm0 sandbox runtime — VM orchestra
 
 ## TLS in Guest Binaries
 
-Guest crates (`guest-agent`, `guest-download`) **must** use system certificate roots (e.g. `rustls-tls-native-roots`), not bundled webpki roots. The host runs a mitmproxy transparent proxy that intercepts HTTPS traffic with its own CA certificate, which is installed into the guest's system certificate store at boot. Using bundled roots would bypass the proxy CA and cause TLS verification failures.
+Guest crates (`guest-agent`, `guest-download`) **must** use system certificate roots, not bundled webpki roots. The host runs a mitmproxy transparent proxy that intercepts HTTPS traffic with its own CA certificate, which is installed into the guest's system certificate store at boot. Using bundled roots would bypass the proxy CA and cause TLS verification failures. As of reqwest 0.13, the `rustls` feature uses `rustls-platform-verifier` which reads from the system certificate store on Linux.
+
+All crates that make HTTP requests should depend on **`reqeast`** (not `reqwest` directly). `reqeast` uses `ring` as the TLS crypto provider instead of `aws-lc-rs`, which is required for `aarch64-unknown-linux-musl` cross-compilation. See [`reqeast/README.md`](reqeast/README.md) for details.
 
 ## Building
 

--- a/crates/ably-subscriber/Cargo.toml
+++ b/crates/ably-subscriber/Cargo.toml
@@ -6,7 +6,7 @@ edition.workspace = true
 [dependencies]
 base64 = { workspace = true }
 futures-util = { workspace = true }
-reqwest = { workspace = true, features = ["json", "rustls-tls-native-roots"] }
+reqeast = { workspace = true }
 rmp-serde = { workspace = true }
 rmpv = { workspace = true }
 serde = { workspace = true, features = ["derive"] }

--- a/crates/ably-subscriber/examples/smoke_test.rs
+++ b/crates/ably-subscriber/examples/smoke_test.rs
@@ -245,7 +245,7 @@ async fn receive_message(
 
 /// Publish a single message via Ably REST API.
 async fn publish_message(
-    client: &reqwest::Client,
+    client: &reqeast::Client,
     rest_host: &str,
     channel: &str,
     auth_header: &str,
@@ -331,7 +331,9 @@ async fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
 
     // --- Publish and verify each test case ---
     let cases = test_cases();
-    let client = reqwest::Client::new();
+    let client = reqeast::builder()
+        .build()
+        .map_err(|e| format!("http client: {e}"))?;
     let mut passed = 0usize;
     let mut failed = 0usize;
 

--- a/crates/ably-subscriber/src/connection.rs
+++ b/crates/ably-subscriber/src/connection.rs
@@ -61,7 +61,7 @@ pub(crate) fn rest_host(realtime_host: &str) -> String {
 
 /// Exchange a TokenRequest for a TokenDetails via Ably's REST API.
 pub(crate) async fn exchange_token(
-    client: &reqwest::Client,
+    client: &reqeast::Client,
     token_request: &crate::TokenRequest,
     host: &str,
 ) -> Result<TokenDetails, Error> {
@@ -293,7 +293,7 @@ pub(crate) struct EventLoopState {
     pub channel_params: Option<HashMap<String, String>>,
     pub realtime_host: String,
     pub rest_host: String,
-    pub http: reqwest::Client,
+    pub http: reqeast::Client,
     pub get_token: Box<dyn Fn() -> TokenFuture + Send + Sync>,
     pub timing: TimingConfig,
     pub token_renewal_failures: u32,

--- a/crates/ably-subscriber/src/subscribe.rs
+++ b/crates/ably-subscriber/src/subscribe.rs
@@ -62,9 +62,7 @@ pub async fn subscribe(config: SubscribeConfig) -> Result<Subscription, Error> {
     let rest = config
         .rest_host
         .unwrap_or_else(|| rest_host(&realtime_host));
-    let http = reqwest::Client::builder()
-        .timeout(timing.connect_timeout)
-        .build()?;
+    let http = reqeast::builder().timeout(timing.connect_timeout).build()?;
 
     // Initial token exchange (with timeout)
     let token_request = tokio::time::timeout(timing.connect_timeout, (config.get_token)())

--- a/crates/ably-subscriber/src/types.rs
+++ b/crates/ably-subscriber/src/types.rs
@@ -194,7 +194,7 @@ pub enum Error {
     WebSocket(Box<tungstenite::Error>),
 
     #[error("Token exchange HTTP error: {0}")]
-    Http(#[from] reqwest::Error),
+    Http(#[from] reqeast::Error),
 
     #[error("MessagePack encode error: {0}")]
     MsgpackEncode(#[from] rmp_serde::encode::Error),

--- a/crates/guest-agent/Cargo.toml
+++ b/crates/guest-agent/Cargo.toml
@@ -8,7 +8,7 @@ base64.workspace = true
 bytes.workspace = true
 guest-common.workspace = true
 libc.workspace = true
-reqwest = { workspace = true, features = ["json", "rustls-tls-native-roots"] }
+reqeast = { workspace = true }
 serde = { workspace = true, features = ["derive"] }
 serde_json.workspace = true
 sha2.workspace = true

--- a/crates/guest-agent/src/http.rs
+++ b/crates/guest-agent/src/http.rs
@@ -5,7 +5,7 @@ use crate::env;
 use crate::error::AgentError;
 use bytes::Bytes;
 use guest_common::log_warn;
-use reqwest::Client;
+use reqeast::Client;
 use serde::Serialize;
 use serde_json::Value;
 use std::sync::LazyLock;
@@ -14,7 +14,7 @@ use std::time::Duration;
 const LOG_TAG: &str = "sandbox:guest-agent";
 
 static HTTP_CLIENT: LazyLock<Client> = LazyLock::new(|| {
-    Client::builder()
+    reqeast::builder()
         .connect_timeout(Duration::from_secs(constants::HTTP_CONNECT_TIMEOUT_SECS))
         .timeout(Duration::from_secs(constants::HTTP_TIMEOUT_SECS))
         .build()

--- a/crates/reqeast/Cargo.toml
+++ b/crates/reqeast/Cargo.toml
@@ -1,0 +1,11 @@
+[package]
+name = "reqeast"
+version = "0.1.0"
+edition.workspace = true
+
+[dependencies]
+reqwest = { workspace = true, features = ["json", "rustls-no-provider"] }
+rustls = { workspace = true }
+
+[lints]
+workspace = true

--- a/crates/reqeast/README.md
+++ b/crates/reqeast/README.md
@@ -1,0 +1,36 @@
+# reqeast
+
+Thin wrapper around [reqwest](https://crates.io/crates/reqwest) that auto-installs the **ring** TLS crypto provider before the first HTTP request.
+
+## Why
+
+reqwest 0.13 switched its default TLS backend from `ring` to `aws-lc-rs`. This causes two problems for us:
+
+1. **Cross-compilation breaks** — `aws-lc-sys` uses glibc-specific fortified functions (`__memcpy_chk`, `__fprintf_chk`) that don't exist in musl libc. Our guest binaries target `aarch64-unknown-linux-musl`, so linking fails.
+
+2. **Manual provider ceremony** — Using reqwest's `rustls-no-provider` feature to opt out of `aws-lc-rs` requires every binary and test to manually call `rustls::crypto::ring::default_provider().install_default()` before any HTTP client is created. Missing this call produces a runtime panic.
+
+`reqeast` solves both by depending on reqwest with `rustls-no-provider` + `ring`, and calling `install_default()` automatically (via `std::sync::Once`) in its `builder()` and `get()` entry points.
+
+## Usage
+
+Use `reqeast::builder()` instead of `reqwest::Client::builder()`:
+
+```rust
+let client = reqeast::builder()
+    .timeout(Duration::from_secs(10))
+    .build()?;
+
+let resp = client.get("https://example.com").send().await?;
+```
+
+Commonly used reqwest types are re-exported (`Client`, `Method`, `StatusCode`, etc.) so downstream crates only need `reqeast` in their `[dependencies]`.
+
+## Background
+
+- [seanmonstar/reqwest#2723](https://github.com/seanmonstar/reqwest/issues/2723) — reqwest 0.13 switched default TLS to rustls + aws-lc-rs (instead of ring).
+- [seanmonstar/reqwest#2630](https://github.com/seanmonstar/reqwest/issues/2630) — Using `rustls-no-provider` requires manual `CryptoProvider::install_default()`; missing it causes a runtime panic.
+- [seanmonstar/reqwest#2136](https://github.com/seanmonstar/reqwest/issues/2136) — Discussion on aws-lc-rs build complexity (needs cmake, nasm) vs ring.
+- [seanmonstar/reqwest#647](https://github.com/seanmonstar/reqwest/issues/647) — Prior art: reqwest in Firecracker + musl — same scenario, same class of problem.
+- [aws/aws-lc-rs#894](https://github.com/aws/aws-lc-rs/issues/894) — `aws-lc-sys` musl build failures (`linux/random.h` missing); fixed but illustrates fragile musl support.
+- [algesten/ureq#751](https://github.com/algesten/ureq/issues/751) — ureq chose to stay on ring; rustls maintainers confirmed "it is fine to stick with ring" for minimal-dependency use cases.

--- a/crates/reqeast/src/lib.rs
+++ b/crates/reqeast/src/lib.rs
@@ -1,0 +1,30 @@
+//! Thin wrapper around [`reqwest`] that auto-installs the `ring` TLS crypto
+//! provider on first use.
+//!
+//! Use [`builder()`] instead of `reqwest::Client::builder()` and [`get()`]
+//! instead of `reqwest::get()` to ensure the provider is installed.
+
+use std::sync::Once;
+
+static INIT: Once = Once::new();
+
+fn ensure_provider() {
+    INIT.call_once(|| {
+        let _ = rustls::crypto::ring::default_provider().install_default();
+    });
+}
+
+/// Create a [`reqwest::ClientBuilder`] with TLS provider auto-initialized.
+pub fn builder() -> reqwest::ClientBuilder {
+    ensure_provider();
+    reqwest::Client::builder()
+}
+
+/// Send a GET request to the given URL (convenience wrapper).
+pub async fn get<U: reqwest::IntoUrl>(url: U) -> reqwest::Result<reqwest::Response> {
+    ensure_provider();
+    reqwest::get(url).await
+}
+
+// Re-export commonly used types.
+pub use reqwest::{Client, Error, IntoUrl, Method, RequestBuilder, Response, StatusCode};

--- a/crates/runner/Cargo.toml
+++ b/crates/runner/Cargo.toml
@@ -20,7 +20,7 @@ tokio = { workspace = true, features = ["rt-multi-thread", "macros", "sync", "si
 tracing = { workspace = true }
 tracing-subscriber = { workspace = true }
 flate2 = { workspace = true }
-reqwest = { workspace = true, features = ["json", "rustls-tls-native-roots"] }
+reqeast = { workspace = true }
 sha2 = { workspace = true }
 tar = { workspace = true }
 tempfile = { workspace = true }

--- a/crates/runner/src/api.rs
+++ b/crates/runner/src/api.rs
@@ -1,4 +1,4 @@
-use reqwest::StatusCode;
+use reqeast::StatusCode;
 use tracing::warn;
 
 use crate::error::{RunnerError, RunnerResult};
@@ -21,7 +21,7 @@ impl ApiClient {
     pub async fn poll(&self, group: &str) -> RunnerResult<Option<Job>> {
         let resp = self
             .http
-            .request(reqwest::Method::POST, "/api/runners/poll", &self.token)
+            .request(reqeast::Method::POST, "/api/runners/poll", &self.token)
             .json(&serde_json::json!({ "group": group }))
             .send()
             .await
@@ -47,7 +47,7 @@ impl ApiClient {
         let path = format!("/api/runners/jobs/{run_id}/claim");
         let resp = self
             .http
-            .request(reqwest::Method::POST, &path, &self.token)
+            .request(reqeast::Method::POST, &path, &self.token)
             .json(&serde_json::json!({}))
             .send()
             .await
@@ -88,7 +88,7 @@ impl ApiClient {
         let resp = self
             .http
             .request(
-                reqwest::Method::POST,
+                reqeast::Method::POST,
                 "/api/webhooks/agent/complete",
                 sandbox_token,
             )
@@ -112,7 +112,7 @@ impl ApiClient {
         let resp = self
             .http
             .request(
-                reqwest::Method::POST,
+                reqeast::Method::POST,
                 "/api/runners/realtime/token",
                 &self.token,
             )

--- a/crates/runner/src/cmd/setup.rs
+++ b/crates/runner/src/cmd/setup.rs
@@ -109,7 +109,7 @@ async fn create_directories(paths: &HomePaths) -> RunnerResult<()> {
 
 /// Stream an HTTP response to a file, computing SHA256 incrementally.
 /// Returns the hex-encoded digest.
-async fn stream_to_file(mut response: reqwest::Response, path: &Path) -> RunnerResult<String> {
+async fn stream_to_file(mut response: reqeast::Response, path: &Path) -> RunnerResult<String> {
     let mut file = tokio::fs::File::create(path)
         .await
         .map_err(|e| RunnerError::Internal(format!("create {}: {e}", path.display())))?;
@@ -135,7 +135,7 @@ async fn stream_to_file(mut response: reqwest::Response, path: &Path) -> RunnerR
 
 /// Download a URL to a temp file. Cleans up on failure. Returns hex SHA256.
 async fn download_to_temp(url: &str, tmp_path: &Path, label: &str) -> RunnerResult<String> {
-    let response = reqwest::get(url)
+    let response = reqeast::get(url)
         .await
         .map_err(|e| RunnerError::Internal(format!("download {label}: {e}")))?;
 

--- a/crates/runner/src/http.rs
+++ b/crates/runner/src/http.rs
@@ -1,7 +1,7 @@
 use std::sync::Arc;
 use std::time::Duration;
 
-use reqwest::Client;
+use reqeast::Client;
 use tracing::info;
 
 use crate::error::{RunnerError, RunnerResult};
@@ -24,7 +24,7 @@ struct Inner {
 
 impl HttpClient {
     pub fn new(api_url: String) -> RunnerResult<Self> {
-        let client = Client::builder()
+        let client = reqeast::builder()
             .timeout(DEFAULT_TIMEOUT)
             .build()
             .map_err(|e| RunnerError::Internal(format!("http client: {e}")))?;
@@ -51,10 +51,10 @@ impl HttpClient {
     /// `path` is appended to the base URL (e.g. `/api/runners/poll`).
     pub fn request(
         &self,
-        method: reqwest::Method,
+        method: reqeast::Method,
         path: &str,
         token: &str,
-    ) -> reqwest::RequestBuilder {
+    ) -> reqeast::RequestBuilder {
         let url = format!("{}{path}", self.inner.api_url);
         let mut req = self.inner.client.request(method, url).bearer_auth(token);
 

--- a/crates/runner/src/network_logs.rs
+++ b/crates/runner/src/network_logs.rs
@@ -1,6 +1,6 @@
 use std::path::Path;
 
-use reqwest::Method;
+use reqeast::Method;
 use serde::{Deserialize, Serialize};
 use tracing::{info, warn};
 use uuid::Uuid;

--- a/crates/runner/src/telemetry.rs
+++ b/crates/runner/src/telemetry.rs
@@ -118,7 +118,7 @@ async fn send_telemetry(http: &HttpClient, run_id: Uuid, sandbox_token: &str, op
 
     let req = http
         .request(
-            reqwest::Method::POST,
+            reqeast::Method::POST,
             "/api/webhooks/agent/telemetry",
             sandbox_token,
         )

